spec/consensus/gossip/crdt.md

```diff
@@ -41,25 +41,19 @@ In other words, the algorithm does not require all messages to be delivered, onl
 One way of looking at the information used by CometBFT nodes is as a distributed tuple space (a set of tuples) to which all nodes contribute.
 Entries are added by validators over possibly many rounds of possibly many heights.
 Each entry has form $\lang h, r, s, v, p \rang$ and corresponds to the message validator node $v$ sent in step $s$ of round $r$ of height $h$; $p$ is a tuple with the message payload.
-To propose a value, the proposer adds it to the tuple space, and to vote for a proposal, a validator adds the vote.
+In the algorithm, whenever a message would be broadcast now a tuple is added to the tuple space.
 
-Each tuple is signed by the validator adding it to the tuple space.
-Since each tuple includes the heigh, round, and step there is no room for forgery (and any attempt to add differing entries for the same heigh, round and step by the same validator will be seen as evidence of misbehavior).[^todo1]
-
-Because nodes are are part of an asynchronous distributed system, individual nodes can only maintain approximations of the tuple space, to which they try to converge.
-There are essentially two ways of making the tuple space converge.
+Because of the asynchronous nature of distributed systems, what a node's view of what is in the tuple space, its **local view**, will differ from the other nodes.
+There are essentially two ways of making converging the local views of nodes.
 
 - **Approach One**: nodes broadcast all the updates they want to perform to all nodes, including themselves.
 If using Reliable Broadcast/the Gossip Communication property, the tuple space will eventually converge to include all broadcast messages.
 - **Approach Two**: nodes periodically compare their approximations with each other, 1-to-1, to identify and correct differences by adding missing entries, using some gossip/anti-entropy protocol.
 
 These approaches work to reach convergence because the updates are commutative regarding the tuple space; each update simply adds an entry to a set.
-From the Tendermint algorithm's point of view, convergence guarantees progress but is not requirement for correctness.[^todo2]
-In other words, nodes observing different approximations of the tuple space may decide at different point in time but cannot violate any correctness guarantees and the eventual convergence of tuple space implies the eventual termination of the algorithm.
-
-[^todo1]: Formalize using [Making CRDTs Byzantine Fault Tolerant](https://martin.kleppmann.com/papers/bft-crdt-papoc22.pdf) as basis; "Many CRDTs, such as Logoot [44] and Treedoc [ 36], assign a unique identifier to each item (e.g. each element in a sequence); the data structure does not allow multiple items with the same ID, since then an ID would be ambiguous.”
+From the Tendermint algorithm's point of view, convergence guarantees progress but is not a requirement for correctness.
```

Comment:
In approach two, I would replace "gossip/anti-entropy" with anti-entropy.

It worths pointing out that approach one requires unbounded space, so in some sense it is not practical.

Reply:
👍 Fixed the second. The first is pointed out in the following section.